plugins/PostProcessingPlugin/scripts/AddCoolingProfile.py

```diff
@@ -0,0 +1,803 @@
+        if by_layer_or_feature == "by_feature" and is_multi_fan:
+            Logger.log("d", "Add Cooling Profile - The MultiFan By Feature section")
+            layer_number = "0"
+            start_index = 1
+            prev_fan = T0_fan
+            this_fan = T0_fan
+            modified_data = ""
+            current_fan_speed = "0"
+            for my_index in range(1, len(data) - 1, 1):
+                layer = data[my_index]
+                if ";LAYER:" + str(the_start_layer) in layer:
+                    start_index = int(my_index) - 1
+                    break
+    #Track the previous tool changes
+            for num in range(1,start_index,1):
+                layer = data[num]
+                lines = layer.split("\n")
+                for line in lines:
+                    if line == "T0":
+                        prev_fan = this_fan
+                        this_fan = T0_fan
+                    elif line == "T1":
+                        prev_fan = this_fan
+                        this_fan = T1_fan
+                    elif line == "T2":
+                        prev_fan = this_fan
+                        this_fan = T2_fan
+                    elif line == "T3":
+                        prev_fan = this_fan
+                        this_fan = T3_fan
+    #Get the current tool.
+            for l_index in range(start_index,start_index + 1,1):
+                layer = data[l_index]
+                lines = layer.split("\n")
+                for line in lines:
+                    if line.startswith("T"):
+                        if line == "T0": this_fan = T0_fan
+                        if line == "T1": this_fan = T1_fan
+                        if line == "T2": this_fan = T2_fan
+                        if line == "T3": this_fan = T3_fan
+                        prev_fan = this_fan
+    #Start to make insertions
+            for l_index in range(start_index+1,len(data)-1,1):
+                layer = data[l_index]
+                lines = layer.split("\n")
+                for line in lines:
+                    if line.startswith("T"):
+                        if line == "T0": this_fan = T0_fan
+                        if line == "T1": this_fan = T1_fan
+                        if line == "T2": this_fan = T2_fan
+                        if line == "T3": this_fan = T3_fan
+                        modified_data += "M106 S0" + prev_fan + "\n" #Turn off the prev fan
+                        modified_data += line + "\n"
+                        modified_data += "M106 S" + str(current_fan_speed) + this_fan + "\n"   #Turn on the current fan
+                        prev_fan = this_fan
+                    if ";LAYER:" in line:
+                        layer_number = str(line.split(":")[1])
+                        modified_data += line + "\n"
+                    if int(layer_number) >= int(the_start_layer):
+                        if ";TYPE:SKIRT" in line:
+                            modified_data += line + "\n"
```

Comment:
Maybe consider making these three _almost_ repeated lines into a (local) function, so you only have to write a call to that function for each 'if' here.

Something like this:

~~~
# Note that this funcion needs to be a local function in order to use `modified_data`, `current_fan_speed` or `this_fan`. If you make it into a normal class function, those parameters will have to be passed as well.
def writeModifiedData(line, fan_sp_type_str):
    modified_data += line + "\n"
    modified_data += fan_sp_type_st + this_fan + "\n"
    current_fan_speed = str(fan_sp_type_str.split("S")[1])

if ";TYPE:SPAM" in line:
    writeModifiedData(line, fan_sp_eggs_spam_spam_eggs)
~~~

(Or, for bonus points: Instead of doing the above, put the `";TYPE:xxx"` and `fan_sp_xxx` in a dictionary `{}` and loop over them with yet another for-loop. -- Though you'd still need to check for the ones that _aren't_ a variation of the same three lines.)

Reply:
Thanks @rburema.
Congratulations on figuring out which version was the latest one.  Every time I try to make an adjustment to the pull request here on Github I've ended up making a mess.

Personally, I print a lot of PETG and finer control of the cooling fan is something that I had wanted for a while.  Without a post processor like this - the fan gets turned on, maybe ramps up (sometimes at speeds that don't get it spinning), finally gets going, and that's it until it gets turned off at the end.  There is no regard for what a print might require in the middle.  I've attempted to address that.
A couple of years ago I made a feature request for this type of post processor but nobody picked up the ball and ran with it.  As a consequence, I decided to do it myself and so you get stuck with me and my very amateurish coding style.

In my defense, I was teaching myself Python as I went along.  I think this Cooling post processor would be an ambitious endeavor even for someone who knows the language.  I am aware that my coding style is decidedly "un-Pythonic".  I cut my programming teeth on an Analog computer and the last programming class I took was Fortran IV in 1973 (we used punch cards to load a program into the IBM360 that took up a very large room).  I see myself as a decent designer, but a poor coder.  I think the bottom line is that the script runs, works, and does what it is supposed to do.  Considering the number of insertions it can make, it's pretty fast too.
I see this as being of special use in industrial situations and printing things like ABS, ASA etc. where some cooling might be required for specific areas, but not in others.  That's why I went with 4 fans and 4 extruders.  In both cases, once you add a second extruder or fan circuit you any additional ones essentially for free.  Many of your comments appear to apply to both of those situations.  I solved the "multi-extruder" and "multi-fan" situation to the best of my ability.

I'm going through your suggestions now.  It seems that most have to do with "methodology" rather than any real changes to the actual logic, and since debugging is a PITA, I likely won't make any real changes to the code.  If you would like to then absolutely be my guest.  I don't want to be flip here, but tightening the code further makes it more difficult for me to read and understand.  There is also the question of repairing the breakage it would cause.  That takes me back to the "debugging is a PITA" thing.
An example is the way I emulate a multi-dimensional array with the "odd/even" values in the fan_list[ ].    I'm actually proud of my workaround since a multi-dimensional array isn't supported by Python, but it isn't like I'm in love with it.  Throwing it away in favor of a second list wouldn't cause me any heartburn.

In regard to the other pull requests - The Alter Z-hops is a new post processor.  There are prints that require Z-hops for a specific range of layers but would not be required for other areas.  It addresses that.
The others are re-writes.  You get a break on those as at least some of the code is written by an actual programmer rather than some random fireman who scribbles.

I did manage to get messages to the user in "AlterZhops".  There are kick-out messages when "Adaptive Layers" is used or when Z-hops simply aren't enabled in the Cura settings.  They come up too late, but they are there.
Search and Replace is a short (65 lines of code) alteration of the existing post processor.  Even I would have trouble screwing that up.

I'll be going back through this and seeing what I can do given your comments (and what I'm actually capable of).  I'll post an update when I get done and can figure out how to do that on Github without adding 17000 commits to the request.

Thanks again,
Greg

Reply:
Hi Greg!

Thanks for your explanation.

My grandfather (mom's dad) and my own dad worked with punch-cards as well, back in the day. (My dad, learned FORTRAN on them when he started at IBM I think). I think we used to have a few at home.

Don't worry about being 'un-Pythonic' _specifically_. I shouldn't generalize, but while there often is _some_ reasoning behind such statements, people using those kinds of monikers too easily are often language-snobs :-)

Which brings me to my first point; Yes, I think these scripts are functional in the sense that they do what is being proposed, and my main issue with them is indeed the methodology.

For us, however, that's not just an aesthetic thing, but connected to maintenance and readability. (Ironically enough, for most modern coders, a version adjusted to my comments would be a lot more readable -- which is indeed probably different for those just starting out with modern languages.)

Given your comments, I think I'm going to have to do most of that rewrite myself. We normally don't do that for rewrites of this size. However, I'll see if I have some leftover time somewhen that I can spend on doing just that. Can't promise anything however! Specifically, it might not happen within the next week (or even two) though. -- Alternatively, since these post-processing scripts are properly encapsulated, I could ask the rest if we can maybe make an exception to certain things this time -- I suspect the answer will be 'no' however.

Now that that's out of the way some more specific things:

> Congratulations on figuring out which version was the latest one. Every time I try to make an adjustment to the pull request here on Github I've ended up making a mess.

I think you can adjust the branch after you made the PR. At least I can for my own PR's.

> Another problem was that I tried to get this into 5.3.1 and so it was a rush job. That isn't a problem now.

For future reference, we almost never add features to patch releases like going from 5.3.0 to 5.3.1 -- Unless you meant 5.3 in general. Even then we can't make promises. (Our track record is not exactly a shining example of punctuality...)

> I'll be going back through this and seeing what I can do given your comments (and what I'm actually capable of). I'll post an update when I get done and can figure out how to do that on Github without adding 17000 commits to the request.

Git has a feature called squash. Some repositories require you to use that before your code is merged, but we're never too worried about that.

In any case, I take it we're not that much of a hurry right now, since it seems 5.4 will take a while as well.

Again, as always, thanks for always helping the community out, one way or the other.

cheers,
Remco.